src/beamlime/constructors/inspectors.py

```diff
@@ -0,0 +1,187 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+# Helper functions for parsing type hints and annotation of a callable object.
+
+from __future__ import annotations
+
+from typing import Any, Callable, Dict, Literal, Type, TypeVar, Union
+
+
+class Empty:
+    ...
+
+
+class InsufficientAnnotationError(Exception):
+    ...
+
+
+class ProductNotFoundError(Exception):
+    ...
+
+
+class UnknownType:
+    ...
+
+
+def validate_annotation(annotation) -> Literal[True]:
+    """
+    Check if the origin of the annotation is not Union.
+
+    If it for later implementation of Union type handling.
+    """
+    from typing import get_origin
+
+    if get_origin(annotation) == Union:
+        raise NotImplementedError("Union annotation is not supported.")
+    return True
+
+
+Product = TypeVar("Product")
+ProductType = Type[Product]
+
+
+def extract_underlying_product_type(product_type: ProductType) -> ProductType:
+    if hasattr(product_type, "__supertype__"):  # NewType
+        return product_type.__supertype__
+    else:
+        return product_type
+
+
+class ProductSpec:
+    """
+    Specification of a product (returned value) of a provider.
+    """
+
+    def __new__(cls, product_type: Union[ProductType, ProductSpec]) -> ProductSpec:
+        if isinstance(product_type, ProductSpec):
+            return product_type
+        else:
+            return super().__new__(ProductSpec)
+
+    def __init__(self, product_type: Union[ProductType, ProductSpec]) -> None:
+        if isinstance(product_type, ProductSpec):
+            ...
+        else:
+            validate_annotation(product_type)
+            self.product_type = product_type
+            self.returned_type = extract_underlying_product_type(product_type)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ProductSpec):
+            raise NotImplementedError(
```

Comment:
Why raise an exception? Typically, Python objects return `False` on type mismatch.

Reply:
Since `ProductSpec` is intended to be used only within the `Provider`,
I wanted to prevent unexpected usage of it as much as possible.
Will it cause any problems later...?

Reply:
> Will it cause any problems later...?

Only really if it is used together with other types. E.g. a function that takes a `ProductSpec` or, say, a function as argument and does an equality check without checking the type.
I doubt that this will be an actual problem. It is just that, AFAIK, the canonical behaviour is to return `False` on type mismatch.

Reply:
Yes, I actually want to strictly forbid it to be used with other types.
I would rather keep it like this and maybe we can change it to return `False` when we want to use it in other places as well.